File: backend/app/api/v1/portfolio.py
"""Portfolio REST API (authenticated, owner-scoped)."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Query

from app.portfolio.dependencies import PortfolioDep
from app.portfolio.types import (
    AllocationBreakdown,
    PortfolioHistory,
    PortfolioSummary,
    SnapshotPoint,
)

router = APIRouter(prefix="/portfolio", tags=["portfolio"])


@router.get("", response_model=PortfolioSummary, summary="Portfolio summary")
async def get_portfolio(portfolio: PortfolioDep) -> PortfolioSummary:
    return await portfolio.summary()


@router.post(
    "/snapshot",
    response_model=SnapshotPoint,
    summary="Create a portfolio snapshot (interval-bucketed, idempotent)",
)
async def create_snapshot(portfolio: PortfolioDep) -> SnapshotPoint:
    """Trigger a snapshot. Also the hook a future scheduler/worker calls."""
    snapshot = await portfolio.create_snapshot()
    return SnapshotPoint(
        snapshot_time=snapshot.snapshot_time,
        equity=snapshot.equity,
        cash=snapshot.cash,
        market_value=snapshot.market_value,
        realized_pnl=snapshot.realized_pnl,
        unrealized_pnl=snapshot.unrealized_pnl,
        total_return_percent=snapshot.total_return_pct,
        daily_pnl=snapshot.daily_pnl,
        exposure_percent=snapshot.exposure_pct,
        position_count=snapshot.position_count,
    )


@router.get("/history", response_model=PortfolioHistory, summary="Portfolio equity history")
async def get_history(
    portfolio: PortfolioDep,
    range: Annotated[str, Query(description="1D|1W|1M|3M|6M|YTD|1Y|ALL")] = "1M",
    start: Annotated[datetime | None, Query(description="Custom start (ISO 8601)")] = None,
    end: Annotated[datetime | None, Query(description="Custom end (ISO 8601)")] = None,
) -> PortfolioHistory:
    return await portfolio.history(range, start=start, end=end)


@router.get("/allocation", response_model=AllocationBreakdown, summary="Portfolio allocation")
async def get_allocation(portfolio: PortfolioDep) -> AllocationBreakdown:
    return await portfolio.allocation()

File: web/app/services/api/adapters/portfolio.ts
import type { PortfolioHistory, PortfolioSnapshot, PortfolioRange, PortfolioSummary } from '~/types/portfolio'
import { iso, num } from './common'
import type { RawPortfolioHistory, RawPortfolioSummary, RawSnapshotPoint } from './raw'

export function toPortfolioSummary(raw: RawPortfolioSummary): PortfolioSummary {
  return {
    portfolio_id: raw.portfolio_id,
    currency: raw.currency,
    cash: num(raw.cash),
    equity: num(raw.equity),
    buying_power: num(raw.buying_power),
    invested: num(raw.invested_amount),
    market_value: num(raw.market_value),
    open_pnl: num(raw.unrealized_pnl),
    realized_pnl: num(raw.realized_pnl),
    unrealized_pnl: num(raw.unrealized_pnl),
    daily_pnl: num(raw.daily_pnl),
    total_pnl: num(raw.total_pnl),
    total_return_pct: num(raw.total_return_percent),
    daily_return_pct: num(raw.daily_return_percent),
    exposure_pct: num(raw.exposure_percent),
    open_positions: raw.position_count,
    updated_at: iso(raw.updated_at),
  }
}

export function toSnapshotPoint(raw: RawSnapshotPoint): PortfolioSnapshot {
  return {
    timestamp: iso(raw.snapshot_time),
    equity: num(raw.equity),
    cash: num(raw.cash),
    invested: num(raw.market_value),
    daily_pnl: num(raw.daily_pnl),
    total_return_pct: num(raw.total_return_percent),
    drawdown_pct: 0,
  }
}

export function toPortfolioHistory(raw: RawPortfolioHistory): PortfolioHistory {
  return {
    range: raw.range as PortfolioRange,
    currency: 'USD',
    points: raw.points.map(toSnapshotPoint),
  }
}
